lib/web/terminal.go

```diff
@@ -92,85 +87,81 @@ type AuthProvider interface {
 
 // NewTerminal creates a web-based terminal based on WebSockets and returns a
 // new TerminalHandler.
-func NewTerminal(ctx context.Context, req TerminalRequest, authProvider AuthProvider, sessCtx *SessionContext) (*TerminalHandler, error) {
-	ctx, span := tracing.DefaultProvider().Tracer("terminal").Start(ctx, "terminal/NewTerminal")
+func NewTerminal(ctx context.Context, cfg TerminalHandlerConfig) (*TerminalHandler, error) {
+	_, span := tracing.DefaultProvider().Tracer("terminal").Start(ctx, "terminal/NewTerminal")
 	defer span.End()
 
 	// Make sure whatever session is requested is a valid session.
-	_, err := session.ParseID(string(req.SessionID))
+	_, err := session.ParseID(cfg.sessionData.ID.String())
 	if err != nil {
 		return nil, trace.BadParameter("sid: invalid session id")
 	}
 
-	if req.Login == "" {
+	if cfg.sessionData.Login == "" {
 		return nil, trace.BadParameter("login: missing login")
 	}
-	if req.Term.W <= 0 || req.Term.H <= 0 {
-		return nil, trace.BadParameter("term: bad term dimensions")
-	}
-
-	servers, err := authProvider.GetNodes(ctx, req.Namespace)
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
 
-	// DELETE IN: 5.0
-	//
-	// All proxies will support lookup by uuid, so host/port lookup
-	// and fallback can be dropped entirely.
-	hostName, hostPort, err := resolveServerHostPort(req.Server, servers)
-	if err != nil {
-		return nil, trace.BadParameter("invalid server name %q: %v", req.Server, err)
+	if cfg.sessionData.ServerID == "" {
+		return nil, trace.BadParameter("server: missing server")
 	}
 
-	var join bool
-	_, err = authProvider.GetSessionTracker(ctx, string(req.SessionID))
-	switch {
-	case trace.IsNotFound(err):
-		join = false
-	case err != nil:
-		return nil, trace.Wrap(err)
-	default:
-		join = true
+	if cfg.term.W <= 0 || cfg.term.H <= 0 ||
+		cfg.term.W >= 4096 || cfg.term.H >= 4096 {
+		return nil, trace.BadParameter("term: bad dimensions(%dx%d)", cfg.term.W, cfg.term.H)
 	}
```

Comment:
Can we refactor this into a `CheckAndSetDefaults` method in `TerminalHandlerConfig`?